cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -422,6 +426,93 @@ impl YaccParser {
+        if let Some(mut j) = self.lookahead_is("<", i) {
+            let mut k = j;
+            let mut lifetimes = HashSet::new();
+            let mut add_lifetime = |k, c: char| {
+                let s = self.src[j..k].trim_start().trim_end().to_string();
+                lifetimes.insert(s);
+                j = k + c.len_utf8();
+                j
+            };
+
+            while k < self.src.len() {
+                let c = self.src[k..].chars().next().unwrap();
+                match c {
+                    '\n' | '\r' => return Err(self.mk_error(YaccParserErrorKind::ReachedEOL, k)),
+                    ',' => {
+                        k = add_lifetime(k, ',');
+                        continue;
+                    }
+                    '>' => {
+                        k = add_lifetime(k, '>');
+                        break;
+                    }
+                    _ => k += c.len_utf8(),
+                }
+            }
+            self.ast.parse_param_lifetimes = Some(lifetimes);
+            i = k;
+        }
+
+        // Next, the '(' pattern : type, ... ')'
+        i = self.parse_ws(i, false)?;
+        if self.lookahead_is("(", i).is_some() {
+            let mut j = i;
+            let mut bindings: Vec<(String, String)> = Vec::new();
+            while j < self.src.len() && self.lookahead_is(")", j).is_none() {
+                let c = self.src[j..].chars().next().unwrap();
+                j += c.len_utf8();
+
+                // Some binding name, or pattern.
+                let k = self.parse_ws(j, false)?;
+                let (k, binding) = self.parse_to_single_colon(k)?;
+                let (k, typ) = self.parse_param_rust_type(k + ':'.len_utf8())?;
+                j = k;
+                bindings.push((binding.trim_end().to_string(), typ));
+            }
+            if !bindings.is_empty() {
+                self.ast.parse_param_bindings = Some(bindings);
+            }
+            i = j;
+        }
+        let (i, _) = self.parse_to_eol(i)?;
+        Ok(self.parse_ws(i, true)?)
+    }
+
+    // Parse a rust type, followed by either a ',' character or an unbalanced ')'
+    // Return the char indice of the trailing character,
+    fn parse_param_rust_type(&mut self, i: usize) -> YaccResult<(usize, String)> {
+        let i = self.parse_ws(i, false)?;
+        let mut j = i;
+        let mut brace_count = 0;
+
+        while j < self.src.len() {
+            let c = self.src[j..].chars().next().unwrap();
+            match c {
+                '\n' | '\r' => return Err(self.mk_error(YaccParserErrorKind::ReachedEOL, j)),
+                ')' | ',' if brace_count == 0 => {
+                    return Ok((j, self.src[i..j].trim_end().to_string()));
+                }
+                '(' | '{' | '[' | '<' => {
+                    brace_count += 1;
+                    j += c.len_utf8();
+                }
+                ')' | '}' | '>' | ']' if brace_count > 0 => {
```

Comment:
I think we can easily catch one source of mistake here by changing this to something like:

```rust
')' | '}' | '>' | ']' => {
    if brace_count == 0 { <error> }
    brace_count -= 1;
    j += c.len_utf8();
}
```

otherwise if someone does close too many brackets, it'll be a bit difficult to work out what happened.

Reply:
Yep, it won't catch the ')' case though.

Reply:
Totally agreed: I'm only inclined to catch the one I suggested because it's so easy it seems almost silly not to do it. There are various other cases (e.g. mismatched brackets) which this code doesn't deal with -- and I'm fine with it not doing so!